Extension/src/LanguageServer/devcmd.ts

```diff
@@ -0,0 +1,238 @@
+}
+
+export function getEffectiveEnvironment(): { [key: string]: string | undefined } {
+    const env = { ...process.env };
+    extensionContext?.environmentVariableCollection.forEach((variable: string, mutator: vscode.EnvironmentVariableMutator) => {
+        if (variable.toLowerCase() === "path") {
+            // Path is special because it has a placeholder to insert the current Path into.
+            env[variable] = resolveVariables(mutator.value);
+        } else {
+            env[variable] = mutator.value;
+        }
+    });
+    return env;
+}
+
+export async function canFindCl(): Promise<boolean> {
+    if (!isWindows) {
+        return false;
+    }
+    const pathOverride = resolveVariables(extensionContext?.environmentVariableCollection.get('Path')?.value);
+    const path = pathOverride ?? process.env['Path'];
+    const found = await whichAsync('cl.exe', path);
+    return isString(found);
+}
+
+export async function setEnvironment(context?: vscode.ExtensionContext) {
+    if (!isWindows) {
+        throw new Error(errorNotWindows);
+    }
+
+    if (!context) {
+        throw new Error(errorNoContext);
+    }
+
+    const vses = await getVSInstallations();
+    if (!vses) {
+        throw new Error(errorNoVSFound);
+    }
+
+    let vs = await chooseVSInstallation(vses);
+    let options: vcvars.Options | undefined;
+    if (!vs) {
+        const compiler = await getAdvancedConfiguration(vses);
+        vs = compiler.vs;
+        options = compiler.options;
+    }
+    const vars = await vscode.window.withProgress({
+        cancellable: false,
+        location: vscode.ProgressLocation.Notification,
+        title: configuringDevEnv
+    }, () => vcvars.getVCVars(vs, options));
+
+    if (!vars || !vars['INCLUDE']) {
+        throw new Error(localize('something.wrong', 'Something went wrong: {0}', JSON.stringify(vars)));
+    }
+
+    const host = vars['VSCMD_ARG_HOST_ARCH'];
+    const target = vars['VSCMD_ARG_TGT_ARCH'];
+    const arch = vcvars.getArchitecture({
+        host: match(host, { 'x86': 'x86', 'x64': 'x64' }) ?? 'x64',
+        target: match(target, { 'x86': 'x86', 'x64': 'x64', 'arm64': 'ARM64', 'arm': 'ARM' }) ?? 'x64'
+    });
+    const settings = new CppSettings();
+
+    context.environmentVariableCollection.clear();
+    for (const key of Object.keys(vars)) {
+        context.environmentVariableCollection.replace(key, vars[key].replace(`%${key}%`, '${env:' + key + '}'));
+    }
+    context.environmentVariableCollection.description = localize('dev.env.for', '{0} developer environment for {1}', arch, vsDisplayNameWithSuffix(vs));
+    context.environmentVariableCollection.persistent = settings.persistVSDeveloperEnvironment;
+
+    return true;
+}
+
+async function getVSInstallations() {
+    const installations = await vswhere.getVSInstallations({
+        all: true,
+        prerelease: true,
+        sort: true,
+        requires: ['Microsoft.VisualStudio.Component.VC.Tools.x86.x64']
```

Comment:
Is this going to work on my (work) ARM64 AI Surface Laptop?  Or should `Microsoft.VisualStudio.Component.VC.Tools.ARM64` be specified there?